internal/thea.go

```diff
@@ -3,323 +3,196 @@ package internal
+func New(config TheaConfig) *theaImpl {
+	log.Emit(logger.DEBUG, "Bootstrapping Thea services using config: %#v\n", config)
 	thea := &theaImpl{
-		cfg:               config,
-		theaCtx:           ctx,
-		theaCtxCancel:     ctxCancel,
-		shutdownWaitGroup: &sync.WaitGroup{},
+		eventBus: event.New(),
+		config:   config,
 	}
 
-	// Inject services
-	thea.UpdateManager = NewUpdateManager(updateFn, thea)
-	thea.ProfileService = NewProfileService(thea)
-	thea.CoreService = NewCoreService(thea)
-	thea.QueueService = NewQueueService(thea)
-	thea.MovieService = NewMovieService(thea)
-
-	// Inject state managers
-	thea.queueMgr = queue.NewProcessorQueue(filepath.Join(cacheDir, THEA_CACHE_FILE_PATH))
-	thea.ffmpegMgr = ffmpeg.NewFfmpegCommander(thea, config.Format)
-	thea.profileMgr = profile.NewProfileList(filepath.Join(configDir, THEA_PROFILE_FILE_PATH))
-	thea.workers = worker.NewWorkerPool()
-
 	return thea
 }
 
-// Start will start Thea by initialising all supporting services/objects and starting
-// the event loops
-func (thea *theaImpl) Start() error {
-	exitChannel := make(chan os.Signal, 1)
-	signal.Notify(exitChannel, os.Interrupt, syscall.SIGTERM)
-
-	log.Emit(logger.DEBUG, "Starting Thea initialisation with config: %#v\n", thea.config())
-
-	defer thea.Stop()
-	if err := thea.initialise(); err != nil {
-		return fmt.Errorf("failed to initialise Thea: %s", err)
+// Run will start all of Thea by bringing up all required services and connections, such as:
+// - Docker services
+// - Stores
+// - Database connection
+// - Service instances
+//
+// This function will not return until Thea is stopped.
+// To stop Thea, the provided context must be cancelled. Errors from which Thea cannot recover
+// will also cause Thea to stop.
+func (thea *theaImpl) Run(parent context.Context) error {
+	thea.dockerManager = docker.NewDockerManager()
+	defer thea.dockerManager.Shutdown(time.Second * 10)
+
+	ctx, cancel := context.WithCancel(parent)
+	crashHandler := func(label string, err error) {
+		log.Emit(logger.FATAL, "Service crash (%s)! %s\n", label, err.Error())
+		cancel()
 	}
 
-	thea.spawnAsyncService(thea.workers.StartWorkers)
-	thea.spawnAsyncService(thea.ffmpegMgr.Start)
-
-	updateTicker := time.NewTicker(THEA_UPDATE_INTERVAL)
-	queueSyncTicker := time.NewTicker(THEA_QUEUE_SYNC_INTERVAL)
-
-	log.Emit(logger.SUCCESS, " --- Thea Startup Complete --- \n")
-
-	for {
-		select {
-		case <-updateTicker.C:
-			thea.SubmitUpdates()
-		case <-queueSyncTicker.C:
-			if err := thea.synchroniseQueue(); err != nil {
-				log.Emit(logger.WARNING, "Failed to synchronise item queue: %s\n", err.Error())
-			}
-		case <-exitChannel:
-			log.Emit(logger.STOP, "Interrupt detected!\n")
-			return nil
-		case <-thea.theaCtx.Done():
-			log.Emit(logger.WARNING, "Context has been cancelled!\n")
-			return nil
-		}
+	log.Emit(logger.NEW, "Initialising Docker services...\n")
+	if err := thea.initialiseDockerServices(thea.config, crashHandler); err != nil {
+		return err
```

Comment:
I find it extremely helpful to add context to every error (ideally with a unique string) since it helps trace the call stack in my log messages. Use `%w` to preserve the internal error within the returned error.

```suggestion
		return fmt.Errorf("could not initialise docker services: %w", err)
```

Reply:
Ooh TIL about %w, thanks!